src/ut/cache_test.cpp

```diff
@@ -1897,4 +1897,159 @@ TEST_F(CacheRequestTest, DissociateImplicitRegistrationSetFromWrongImpi)
   EXPECT_TRUE(log.contains("not all the provided IMPIs are associated with the IMPU"));
 }
 
+//
+// Tests for listing IMPUs in the cache.
+//
+
+// Helper class to save off the IMPUs retrieved from the database.
+class ListImpusRecorder : public ResultRecorderInterface
+{
+public:
+  void save(CassandraStore::Operation* op)
+  {
+    impus = dynamic_cast<Cache::ListImpus*>(op)->get_impus_ref();
+  }
+  std::vector<std::string> impus;
+};
+
+
+// Tests listing IMPUs when there aren't any in the database.
+TEST_F(CacheRequestTest, ListImpusNoResults)
+{
+  ListImpusRecorder rec;
+  RecordingTransaction* trx = make_rec_trx(&rec);
+  CassandraStore::Operation* op = _cache.create_ListImpus();
+
+  std::vector<cass::KeySlice> result;
+  std::vector<std::string> requested_columns = {"_exists", "is_registered"};
+  EXPECT_CALL(_client,
+              get_range_slices(_,
+                               ColumnPathForTable("impu"),
+                               SpecificColumns(requested_columns),
+                               KeysInRange("", ""),
+                               _))
+    .WillOnce(SetArgReferee<0>(result));
+  EXPECT_CALL(*trx, on_success(_))
+    .WillOnce(Invoke(trx, &RecordingTransaction::record_result));
+
+  execute_trx(op, trx);
+
+  EXPECT_TRUE(rec.impus.empty());
+}
+
+// Tests listing IMPUs when there there is on IMPU, that has been dynamically
+// created rather than configured.
+TEST_F(CacheRequestTest, ListImpusOneResultRegistered)
+{
+  ListImpusRecorder rec;
+  RecordingTransaction* trx = make_rec_trx(&rec);
+  CassandraStore::Operation* op = _cache.create_ListImpus();
+
+  std::vector<cass::KeySlice> result(1);
+  result[0].key = "Kermit";
+  // Set the "is_registered" column to make the subscriber appear registered.
+  make_slice(result[0].columns, {{"is_registered", "\x01"}});
+
+  EXPECT_CALL(_client, get_range_slices(_, _, _, _, _))
+    .WillOnce(SetArgReferee<0>(result));
+  EXPECT_CALL(*trx, on_success(_))
+    .WillOnce(Invoke(trx, &RecordingTransaction::record_result));
+
+  execute_trx(op, trx);
+
+  EXPECT_EQ(1, rec.impus.size());
+  EXPECT_EQ("Kermit", rec.impus[0]);
+}
+
+// Tests listing IMPUs when there there is on IMPU, that has been configured but
+// is NOT registered.
+TEST_F(CacheRequestTest, ListImpusOneResultConfigured)
+{
+  ListImpusRecorder rec;
+  RecordingTransaction* trx = make_rec_trx(&rec);
+  CassandraStore::Operation* op = _cache.create_ListImpus();
+
+  std::vector<cass::KeySlice> result(1);
+  result[0].key = "Kermit";
+  // Set the _exists column to show the subscriber is configured.
+  make_slice(result[0].columns, {{"_exists", ""}});
```

Comment:
The distinction between this test and the previous one is probably not relevant any more. Anyway, I'm going to stop commenting on this and let you sort it out.